bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/HydrawiseCloudHandler.java

```diff
@@ -0,0 +1,246 @@
+            updateGroupState(group, CHANNEL_SENSOR_OFFTIMER, new DecimalType(sensor.getOfftimer()));
+            updateGroupState(group, CHANNEL_SENSOR_TIMER, new DecimalType(sensor.getTimer()));
+        });
+    }
+
+    private void updateForecast(StatusScheduleResponse status) {
+        int i = 1;
+        for (Forecast forecast : status.getForecast()) {
+            String group = "forecast" + (i++);
+            updateGroupState(group, CHANNEL_FORECAST_CONDITIONS, new StringType(forecast.getConditions()));
+            updateGroupState(group, CHANNEL_FORECAST_DAY, new StringType(forecast.getDay()));
+            updateGroupState(group, CHANNEL_FORECAST_HUMIDITY, new DecimalType(forecast.getHumidity()));
+            updateTemperature(forecast.getTempHi(), group, CHANNEL_FORECAST_TEMPERATURE_HIGH);
+            updateTemperature(forecast.getTempLo(), group, CHANNEL_FORECAST_TEMPERATURE_LOW);
+            updateWindspeed(forecast.getWind(), group, CHANNEL_FORECAST_WIND);
+        }
+    }
+
+    private void updateTemperature(String tempString, String group, String channel) {
+        Matcher matcher = TEMPERATURE_PATTERN.matcher(tempString);
+        if (matcher.matches()) {
+            try {
+                updateGroupState(group, channel, new QuantityType<Temperature>(Float.parseFloat(matcher.group(1)),
+                        "C".equals(matcher.group(2)) ? SIUnits.CELSIUS : ImperialUnits.FAHRENHEIT));
+            } catch (NumberFormatException e) {
+                logger.debug("Could not parse temperature string {} ", tempString);
+            }
+        }
+    }
+
+    private void updateWindspeed(String windString, String group, String channel) {
+        Matcher matcher = WIND_SPEED_PATTERN.matcher(windString);
+        if (matcher.matches()) {
+            try {
+                updateGroupState(group, channel, new QuantityType<Speed>(Integer.parseInt(matcher.group(1)),
+                        "kph".equals(matcher.group(2)) ? SIUnits.KILOMETRE_PER_HOUR : ImperialUnits.MILES_PER_HOUR));
+            } catch (NumberFormatException e) {
+                logger.debug("Could not parse wind string {} ", windString);
+            }
+        }
+    }
+
+    private void updateControllerProperties(Controller controller) {
+        getThing().setProperty(PROPERTY_CONTROLLER_ID, String.valueOf(controller.getControllerId()));
+        getThing().setProperty(PROPERTY_NAME, controller.getName());
+        getThing().setProperty(PROPERTY_DESCRIPTION, controller.getDescription());
+        getThing().setProperty(PROPERTY_LOCATION, controller.getLatitude() + "," + controller.getLongitude());
+        getThing().setProperty(PROPERTY_ADDRESS, controller.getAddress());
+    }
+
+    private @Nullable Controller getController(int controllerId, List<Controller> controllers) {
+        try {
+            return controllers.stream().filter(c -> controllerId == c.getControllerId().intValue()).findAny().get();
```

Comment:
Does this work also (So you don't need to catch the exception)?
```suggestion
            return controllers.stream().filter(c -> controllerId == c.getControllerId().intValue()).findAny().orElse(null);
```

Reply:
That would be nicer, but the compiler complains:
`Null type mismatch: required '@NonNull Controller' but the provided value is null`
And i don't see a way to make it happy.

Reply:
I indeed recently saw as well that you could not `null `to `orElse` (because of our settings?), which kinda pissed me off :-)

Instead, I would then prefer just an if (isPresent), throwing and handling the exception is more expensive and worst practice.